src/contracts/PresentProtocol.sol

```diff
@@ -8,65 +8,152 @@ import "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
 import "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
 import "@openzeppelin/contracts/utils/introspection/ERC165Checker.sol";
 
-contract PresentProtocol is ERC721, ERC721Holder, ERC1155Holder, Ownable {
-    bytes4 constant _INTERFACE_ID_ERC721 = 0x80ac58cd;
-    bytes4 constant _INTERFACE_ID_ERC1155 = 0xd9b67a26;
+import "./Base64.sol";
+import "./IPresentProtocol.sol";
 
+contract PresentProtocol is IPresentProtocol, ERC721, ERC721Holder, ERC1155Holder, Ownable {
+    using Strings for uint160;
+    using Strings for uint256;
+    bytes4 constant _INTERFACE_ID_ERC721  = 0x80ac58cd;
+    bytes4 constant _INTERFACE_ID_ERC1155 = 0xd9b67a26;
     string  public baseURI;
     uint256 public currentId;
-    mapping(uint256 => Present) public presents;
+    uint256 public fee;
+    mapping(uint256 => bytes) public presents;
+    mapping(uint256 => string) internal messages;
 
-    struct Present {
-        address nftContract;
-        uint256 tokenId;
-    }
+    constructor() ERC721("Present Protocol", "GIFT") {}
 
-    error InvalidContract();
-    error NotAuthorized();
+    function wrap(bytes calldata _gift, address _to, string calldata _message) external payable {
+        if (msg.value != fee) revert InvalidPayment();
 
-    event Wrapped(address indexed _nftContract, uint256 indexed _tokenId, address indexed _gifter, address _receiver, uint256 _presentId);
-    event Unwrapped(address indexed _nftContract, uint256 indexed _tokenId, address indexed _receiver, uint256 _presentId);
+        bytes memory from = abi.encode(msg.sender);
+        bytes memory present = abi.encodePacked(from, _gift);
+        (, address nftContract, uint256 tokenId, ) = _decode(present);
```

Comment:
honestly, it's been 4 months since I've looked at this code. I don't even know what the reasoning for it is lol

I should've added some natspec 😅

Reply:
ahhh I think it's because gift is already encoded when passed in (using the `encode` view function), but we want to add `msg.sender` to it. so we do that first and then decode the nft contract and tokenId to perform the transfer

Reply:
ya could be fixed with the struct thing but if its easier for frontend I think its fine